src/emulator/storage/files.ts

```diff
@@ -149,14 +245,47 @@ export class StorageLayer {
    * @throws {ForbiddenError} if request is unauthorized
    */
   public async getObject(request: GetObjectRequest): Promise<GetObjectResponse> {
+    let authorized = true;
     const metadata = this.getMetadata(request.bucketId, request.decodedObjectId);
 
     // If a valid download token is present, skip Firebase Rules auth. Mainly used by the js sdk.
     const hasValidDownloadToken = (metadata?.downloadTokens || []).includes(
       request.downloadToken ?? ""
     );
-    let authorized = hasValidDownloadToken;
-    if (!authorized) {
+
+    let checkAuth = false;
+
+    if (!hasValidDownloadToken) {
+      if (request.signedUrl?.signature) {
+        const start = validateSignedUrlAndReturnStartTs(request.signedUrl);
+
+        const changeInTime = request.signedUrl.ttlSeconds! * SECONDS_TO_MS_FACTOR;
+        const end = start + changeInTime;
+        const now = convertDateToMS(getCurrentDate());
+
+        const isLive = now >= start && now < end;
+        if (!isLive) {
+          throw new BadRequestError("Url has Expired");
+        }
+
+        const unsignedUrl = createUnsignedUrl({
+          bucketId: request.bucketId,
+          decodedObjectId: request.decodedObjectId,
+          url: request.signedUrl.base as string,
+          usableSeconds: request.signedUrl.usableSeconds!,
+          ttlSeconds: request.signedUrl.ttlSeconds!,
+        });
+
+        const isCorrect = createSignature(unsignedUrl) === request.signedUrl.signature;
+        if (!isCorrect) {
+          throw new ForbiddenError("Invalid Url");
+        }
+      } else {
+        checkAuth = true;
+      }
+    }
+
+    if (checkAuth) {
```

Comment:
We can get rid of this variable and just do. We should check auth by default, and then make the exceptions clear.

```
if (!(hasValidDownloadToken || hasValidSignedUrl)) {
  ...
}
```

Reply:
checkAuth just makes it more obvious of why we are making this check. With the introduction of one variable, we make it easy for a user to see why we would checkAuth following when it gets set to true.

Reply:
But if you think your way is better we can hop on a call to talk about it more